Update whitespace handling

mattgarrish opened the pull request:
As discussed on the 2021-02-26 telecon, this PR adds a cross-reference to infra spec definition for stripping leading and trailing whitespace from metadata elements in the reading system spec definition.

I've also created a new section in the core spec under the package metadata section for common metadata value requirements. These are the only elements that accept text values, so I've moved the floating paragraph that used be at the start of the package doc definition into it. I've also added a single statement to cover the requirement for at least one character after whitespace stripping so we don't have to maintain these statements in every definition.

This only partially solves #1295 and #1528 since we haven't yet addressed whitespace collapsing within elements.


- Reading Systems [preview](https://cdn.statically.io/gh/w3c/epub-specs/editorial/issue-1295/epub33/rs/index.html)
- Reading Systems [diff](https://services.w3.org/htmldiff?doc1=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://w3c.github.io/epub-specs/epub33/rs/index.html&doc2=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://cdn.statically.io/gh/w3c/epub-specs/editorial/issue-1295/epub33/rs/index.html)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/1539.html" title="Last updated on Mar 1, 2021, 4:34 PM UTC (8ee4fd0)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/1539/7a60ccf...8ee4fd0.html" title="Last updated on Mar 1, 2021, 4:34 PM UTC (8ee4fd0)">Diff</a>

mattgarrish committed:
add cross-ref to infra spec definition for stripping leading and trailing whitespace from metadata elements;
create new section for common metadata value requirements to avoid proliferation of trimming requirement and better place para on element values

rdeltour commented on epub33/core/index.html:
```diff
@@ -1560,6 +1557,18 @@ <h6>The <code>metadata</code> Element</h6>
+
+							<p>When a child element of the <a href="#sec-metadata-elem"><code>metadata</code>
+									element</a> has mandatory text content, that content is referred to as the value of
+								the element in the explanatory descriptions.</p>
```
The term being defined could be marked up to stand out and so that it can be referenced from the other spec?

Also, "value of the element" can be confusing. A better terms might be "value of the metadata expression" or something along these lines (
("element" is markup terminology, here we're defining metadata terminology).

Finally, it would perhaps be easier to define the "value" in the general case (empty or not), and then state case-by-case when the value must be non-empty. In the current wording, the term is only defined when the value is mandatory (i.e. not empty).

mattgarrish replied:
I think what you're proposing might make the text more confusing. The element definitions talk in places about their "value". The meta element has a value, for example, but to get to a "metadata expression" you need to also be talking about what is expressed in its property attribute (or the inherent semantics of the DCMES elements).

Looking again at this though, what might help is to be even more specific. We know exactly what elements this applies to - the DCMES elements and meta - so we can list those.

I'm fine with setting value up as a definition that we can link to, though. That makes sense.

> Finally, it would perhaps be easier to define the "value" in the general case (empty or not)

Note completely sure what you mean about defining the "value"? The elements already have a content model of text in their definitions. Are you suggesting we try to further define what that means?

rdeltour replied:
> I think what you're proposing might make the text more confusing. The element definitions talk in places about their "value". The meta element has a value, for example, but to get to a "metadata expression" you need to also be talking about what is expressed in its property attribute (or the inherent semantics of the DCMES elements).

OK fair enough. I'm not too hung-up on _not_ using "element value", as long as the term is a proper definition that can be referenced. In fact, I came here from the RS spec preview, when I found "element value" and was confused by that terminology which didn't sound familiar (from HTML or XML). So it's probably OK if uses of the term can be linked to the definition.

> Note completely sure what you mean about defining the "value"? The elements already have a content model of text in their definitions. Are you suggesting we try to further define what that means?

I meant that the "value" of an element is not defined by XML, so we should be more explicit about what we mean here. DOM defines a [`nodeValue`](https://dom.spec.whatwg.org/#dom-node-nodevalue) attribute for nodes, which for elements is unambiguously defined as the [descendant text context](https://dom.spec.whatwg.org/#concept-descendant-text-content). I think we should use a similar unambiguous definition of what is exactly meant by "value" here.

The current text says:
> "When a child element of the metadata element has mandatory text content, that content is referred to as the value of the element in the explanatory descriptions."

As all the elements require a non-empty value, why make a case ("when") for mandatory text content when defining what we refer to when we say "the value of the element"?

What I suggest is something along these lines:

- add a definition for "the value of the element" (not my favorite term, but OK), based on a concept similar to DOM's _descendant text content_, without the "when".
- add the MUST not be empty statement for that term just defined, based on INFRA.

So a lot like what is already there, but more explicit :)

mattgarrish replied:
> I'm not too hung-up on _not_ using "element value"

For context, this complaint came a couple of revisions back that we were using "value" for element text content when the XML spec doesn't define such a concept. The prime reason I want to stick with it is that changing to something else like "metadata expression" means we probably also have to pick through the whole spec and try to rewrite the references so that they also make sense about the text content being the expression. This made people happy at the time, so I'd rather not try to reinvent it, not that I necessarily love the terminology either. 😄 

> * add a definition for "the value of the element" (not my favorite term, but OK), based on a concept similar to DOM's _descendant text content_, without the "when".

That's probably too complex as EPUB doesn't have element nesting. Even the DOM definition of child text content is confusing in our context, as there's only one instance of child text content so nothing to concatenate.

An element has text content and that's it. I don't know what else we can compare that with, but I don't have time today to skim the usual specs for definitions.

I did make a few changes before signing off last night, like ditching the "when", adding a specific list of elements this applies to, and changing the wording to non-empty. If you have any thoughts on what we can do with text content, though, let me know.

mattgarrish replied:
> An element has text content and that's it.

I wasn't thinking you could put anything but text in an element, but I suppose comments, processing instructions and even cdata blocks are all valid. Epubcheck isn't complaining when I add them.

Given that, the child text content definition is correct since there could be a concatenation in some extreme cases.

mattgarrish replied:
I've made some more changes to add the child text content definition and infra whitespace handling to the core spec, and also removed the exception for meta properties defining their own whitespace handling from the RS. Let me know if you're happier with the sections now (I'm going to close these issues just because they reference out of date prose now).

rdeltour replied:
> Let me know if you're happier with the sections now

Looks better to me now! Thanks for the changes (and for bearing with my nitpicking 😅).

rdeltour commented on epub33/core/index.html:
```diff
@@ -1560,6 +1557,18 @@ <h6>The <code>metadata</code> Element</h6>
+
+							<p>These values MUST be at least one character in length after <a
+									href="https://www.w3.org/TR/epub-rs-33/#conf-meta-ws">stripping leading and trailing
+									ASCII whitespace</a> [[!EPUB-RS-33]].</p>
```
I would put the "stripping leading and trailing" in the definition of the value, and put "value MUST be non-empty" statements case-by-case in the metadata property definitions.

If all metadata expressions mandate non-empty values, then the above definition of value can be simplified (no "When"), and the "value MUST be non-empty statement" can be defined for all.

Also, shouldn't "stripping leading and trailing" refer to INFRA rather than EPUB-RS here?

mattgarrish replied:
> If all metadata expressions mandate non-empty values, then the above definition of value can be simplified (no "When"), and the "value MUST be non-empty statement" can be defined for all.

Right, this is what being more specific might help with. There is a finite set of elements and they all require a non-empty value. That's why I'd rather define this once than list it for each. It's always been cumbersome.

> Also, shouldn't "stripping leading and trailing" refer to INFRA rather than EPUB-RS here?

I considered that, but it's technically the RS that does the stripping. If you want to find out more, the first hop should be to the RS spec.

rdeltour replied:
> > Also, shouldn't "stripping leading and trailing" refer to INFRA rather than EPUB-RS here?
> 
> I considered that, but it's technically the RS that does the stripping. I

Even if the RS will effectively do the stripping at some point, the stripping is a fundamental part of the MUST statement (or of the definition of "value", alternatively).

As an example (out of thin air 😇😁),  a checker would need to do the stripping too, it shouldn't have to go through the RS spec indirection to interpret this MUST statement 😊.

mattgarrish replied:
> a checker would need to do the stripping too, it shouldn't have to go through the RS spec indirection to interpret this MUST statement

I'd argue it does, though, as the stripping is conditional.

For DC elements its required, but there's a loophole for the meta element that properties can override the default behaviour. This is underspecified, in my opinion, as I believe the intention was only that properties _we_ define can override, not that any property defined in any vocabulary can change our rules. That would be a nightmare of inconsistency to deal with.

But given that I'm not aware of any properties we define that change the whitespace handling rules, I wonder if we can just do away with that rule entirely?

I'd then be more at ease with each spec separately referencing across to infra.

rdeltour commented on epub33/rs/index.html:
```diff
@@ -341,13 +341,15 @@ <h4>Metadata</h4>
+								algorithm, Reading Systems MUST <a
+									href="https://infra.spec.whatwg.org/#strip-leading-and-trailing-ascii-whitespace"
+									>strip leading and trailing ASCII whitespace</a> [[!Infra]] from <code>meta</code>
+								element values before further processing them.</p>
```
I was wondering what "element values" meant (when reading the preview), before seeing that this PR defined that term in the core spec. I suggest (see comment above) a more specific term is used (e.g. "metadata value" or "metadata expression value"), and that its usage here points to its definition in the Core spec.

rdeltour reviewed (changes requested):
Looks good generally, but I suggest the "value" concept could be better defined, be referenceable, and use less markupy and more metadatay terminology (e.g. "metadata value" instead of "element value"). See inline comments.

mattgarrish committed:
add dfn to value and link instances to it;
clarify the specific elements that have text values

mattgarrish committed:
add reference to child text content for value of elements;
add link to infra spec in core spec for whitespace handling;
remove exception that properties can define their own whitespace handling rules

mattgarrish committed:
Merge branch 'main' into editorial/issue-1295
